### ISSUE TITLE
DIskFileUpload misbehaves if addContent is called multiple times

### ISSUE THREAD

**Issue by zentol** [netty/netty#10425]
### Expected behavior

Calling `DiskFileUpload#addContent()` multiple times appends the contents of the ByteBufs to the file.

Furthermore, the ByteBuf readerIndex is advanced by the number of bytes written.

### Actual behavior

On each call, the content of the ByteBuf is written to the start of the file. This renders the DiskFileUpload unusable for larger files.
The ByteBuf readerindex is not advanced.

### Steps to reproduce

See below.

### Minimal yet complete reproducer code (or URL to code)

(Modified `io.netty.handler.codec.http.multipart.DiskFileUploadTest`)

```
@Test
public void testAddContentFromByteBuf() throws Exception {
    testAddContentFromByteBuf0(false);
}

@Test
public void testAddContentFromCompositeByteBuf() throws Exception {
    testAddContentFromByteBuf0(true);
}

private static void testAddContentFromByteBuf0(boolean composite) throws Exception {
    DiskFileUpload f1 = new DiskFileUpload("file3", "file3", "application/json", null, null, 0);
    try {
        byte[] bytes = new byte[4096];
        PlatformDependent.threadLocalRandom().nextBytes(bytes);

        List<ByteBuf> buffers = Arrays.asList(createBuffer(bytes, composite), createBuffer(bytes, composite));

        for (int i = 0; i < buffers.size(); i++) {
            f1.addContent(buffers.get(i), i == buffers.size() - 1);
        }

        ByteBuf buf = f1.getByteBuf();
        assertEquals(buf.readerIndex(), 0);
        assertEquals(buf.writerIndex(), bytes.length * 2);

        for (ByteBuf buffer : buffers) {
            assertEquals(buffer.readerIndex(), bytes.length);
        }
        assertArrayEquals(bytes, ByteBufUtil.getBytes(buf, 0, bytes.length));
        assertArrayEquals(bytes, ByteBufUtil.getBytes(buf, bytes.length, bytes.length));
    } finally {
        //release the ByteBuf
        f1.delete();
    }
}

private static ByteBuf createBuffer(byte[] bytes, boolean composite) {
    final ByteBuf buffer;

    if (composite) {
        buffer = Unpooled.compositeBuffer()
                .addComponent(true, Unpooled.wrappedBuffer(bytes, 0 , bytes.length / 2))
                .addComponent(true, Unpooled.wrappedBuffer(bytes, bytes.length / 2, bytes.length / 2));
    } else {
        buffer = Unpooled.wrappedBuffer(bytes);
    }

    return buffer;
}
```

### Netty version

4.1.50
(caused by https://github.com/netty/netty/pull/10360)

### JVM version (e.g. `java -version`)

openjdk version "1.8.0_212"
OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_212-b03)
OpenJDK 64-Bit Server VM (AdoptOpenJDK)(build 25.212-b03, mixed mode)

### OS version (e.g. `uname -a`)

Windows 10

### Proposal

Modify `DiskFileUpload` as follows:

```
@@ -164,7 +164,8 @@ public abstract class AbstractDiskHttpData extends AbstractHttpData {
                     RandomAccessFile accessFile = new RandomAccessFile(file, "rw");
                     fileChannel = accessFile.getChannel();
                 }
-                buffer.getBytes(buffer.readerIndex(), fileChannel, fileChannel.position(), localsize);
+                written = buffer.getBytes(buffer.readerIndex(), fileChannel, fileChannel.position(), localsize);
+                fileChannel.position(fileChannel.position() + written);
                 size += localsize;
                 buffer.readerIndex(buffer.readerIndex() + written);
             } finally {
```

**Comment by fredericBregier** [comment]
@normanmaurer It might be the root cause of bad behavior in reading (get()) later on.
But I believe that the proposal is not enough: one might need to do a `while loop` around the two added lines in order to ensure all the buffer is written (and therefore size will be coherent). Something like:

      -                buffer.getBytes(buffer.readerIndex(), fileChannel, fileChannel.position(), localsize);
      +                int written = 0;
      +                int toWrite = localsize;
      +                while (toWrite > 0) {
      +                    written = buffer.getBytes(buffer.readerIndex(), fileChannel, fileChannel.position(), toWrite);
      +                    if (written < 0) {
      +                       break;
      +                    }
      +                    fileChannel.position(fileChannel.position() + written);
      +                    buffer.setReaderIndex(buffer.readerIndex() + written);
      +                    toWrite -= written;
      +                }
                    size += localsize;

Note: sorry for the web through discussion, I do not have better access right now...

**Comment by fredericBregier** [comment]
Probably to be closed since #10449 #10453
